zipline/pipeline/engine.py

```diff
@@ -478,25 +487,45 @@ def compute_chunk(self, graph, dates, assets, initial_workspace):
         results : dict
             Dictionary mapping requested results to outputs.
         """
-        self._validate_compute_chunk_params(dates, assets, initial_workspace)
-        get_loader = self.get_loader
+        self._validate_compute_chunk_params(
+            graph, dates, sids, initial_workspace,
+        )
+        get_loader = self._get_loader
 
         # Copy the supplied initial workspace so we don't mutate it in place.
         workspace = initial_workspace.copy()
         refcounts = graph.initial_refcounts(workspace)
         execution_order = graph.execution_order(refcounts)
-
-        # If loadable terms share the same loader and extra_rows, load them all
-        # together.
-        loadable_terms = graph.loadable_terms
-        loader_group_key = juxt(get_loader, getitem(graph.extra_rows))
+        domain = graph.domain
+
+        # Many loaders can fetch data more efficiently if we ask them to
+        # retrieve all their inputs at once. For example, a loader backed by a
+        # SQL database can fetch multiple columns from the database in a single
+        # query.
+        #
+        # To enable these loaders to fetch their data efficiently, we group
+        # together requests for LoadableTerms if they are provided by the same
+        # loader and they require the same number of extra rows.
+        #
+        # The extra rows condition is a simplification: we don't currently have
+        # a mechanism for asking a loader to fetch different windows of data
+        # for different terms, so we only batch requests together when they're
+        # going to produce data for the same set of dates. That may change in
+        # the future if we find a loader that can still benefit significantly
+        # from batching unequal-length requests.
+        def loader_group_key(term):
+            loader = get_loader(term)
+            extra_rows = graph.extra_rows[term]
+            return loader, extra_rows
+
+        # Only produce loader groups for the terms we expect to load.  This
+        # ensures that we can run pipelines for graphs where we don't have a
+        # loader registered for an atomic term if all the dependencies of that
+        # term were supplied in the initial workspace.
+        will_be_loaded = graph.loadable_terms - viewkeys(workspace)
         loader_groups = groupby(
             loader_group_key,
-            # Only produce loader groups for the terms we expect to load.  This
```

Comment:
was this comment wrong before?

Reply:
Yeah, I think previously we unconditionally called `loader_group_key` on loadable terms even if they were pre-supplied in the workspace.